core/orchestrator.go

```diff
@@ -517,49 +514,58 @@ type RemoteTranscoderManager struct {
 func (rtm *RemoteTranscoderManager) Register(transcoder *RemoteTranscoder) {
 	rtm.RTmutex.Lock()
 	defer rtm.RTmutex.Unlock()
+
 	rtm.liveTranscoders[transcoder.stream] = transcoder
 
 	for i := 0; i < transcoder.Capacity(); i++ {
-		rtm.AppendTranscoder(transcoder)
-	}
-}
-
-func (rtm *RemoteTranscoderManager) AppendTranscoder(transcoder *RemoteTranscoder) {
-	// rtm.RTmutex.Lock()
-	// defer rtm.RTmutex.Unlock()
-	if _, ok := rtm.liveTranscoders[transcoder.stream]; ok {
 		rtm.remoteTranscoders = append(rtm.remoteTranscoders, transcoder)
 	}
-	return
 }
 
 func (rtm *RemoteTranscoderManager) Unregister(t *RemoteTranscoder) {
 	rtm.RTmutex.Lock()
+	delete(rtm.liveTranscoders, t.stream)
 	defer rtm.RTmutex.Unlock()
-	for _, liveT := range rtm.liveTranscoders {
-		if liveT.stream == t.stream {
-			delete(rtm.liveTranscoders, t.stream)
-		}
+}
+
+func (rtm *RemoteTranscoderManager) SelectTranscoder() *RemoteTranscoder {
+	rtm.RTmutex.Lock()
+	defer rtm.RTmutex.Unlock()
+
+	checkTranscoders := func(rtm *RemoteTranscoderManager) bool {
+		return len(rtm.remoteTranscoders) > 0
 	}
 
-	var remoteTranscoders []*RemoteTranscoder
-	for _, v := range rtm.remoteTranscoders {
-		if v.stream == t.stream {
-			continue
+	for checkTranscoders(rtm) {
+		currentTranscoder, remoteTranscoders := rtm.remoteTranscoders[0], rtm.remoteTranscoders[1:]
```

Comment:
Should we take the same "stack based" approach as multi-O to reduce some of the load here, especially with appends? That is, select from the end rather than the beginning (and append back on the end) so the backing store doesn't need to be continuously reallocated.

Reply:
Sure we can do that. But is it a concern if some of the transcoders end up never getting any work, though they are comparably good Ts? Sure, if one T is getting all the work, that T probably has a higher capacity, can do transcoding faster, and in a free market, should get as many segments as it can handle ... and sure, T's could move onto working with a different O if they're not getting too much work. But, is that going to promote large players from taking all work, and smaller players to flounder? I imagine a lot of O's might run their own transcoders, so it wouldn't matter for them.

Reply:
At this point, the economics / incentives of Ts isn't really a consideration here.

There is something to be said about spreading the load evenly among Ts - currently, the way we construct lists on registration leads to Ts being "clumped" together. Whether we use a stack or a queue wouldn't change this clumping behavior. There are ways we can optimize this, but I don't think it's something we need to address immediately.

Reply:
Got it, we can address later then!